New feature: Add Feed-specific 'Delete After Playback' override option

udif opened the pull request:
Let me repeat this again:
DO NOT PULL THIS AS-IS. THIS IS FOR DISCUSSION ONLY.

One of the features I still miss in AntennaPod is the ability to control the Auto Delete option *per feed*.
Some feeds are throwaway - once I hear them, I want them to be automatically delete.
Other feeds are precious and I want to keep the chapters I've downloaded.

This pull request adds a new 'Delete After Playback' spinner to the FeedInfoActivity. It has 3 states:
Default - use the global setting
Yes - delete after playback, regardless of the global settings
No - Do not delete after playback, regardless of the global settings

This is my first Android work, and I have a VERY shallow understanding of Java (I come from C/C++), so this pull request is based on googling stackoverflow and the SDK documentation.

I know for a fact that this code does NOT work yet. I ran it through the emulator and I see that my settings is not saved between invocations. I'm sending this to solicit feedback and to learn `android programming.

I also hope I didn't do anything stupid with git. There were a few commits on the main repository between the time I forked and the time I merged my changes back to create the pull request.

ChaoticMind commented:
re: git -- if there are changes in the main repo, you should [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) your changes on top of master instead of merging master into your PR

TomHennen commented:
Can you include screenshots of the feature? That would make it easier for people to evaluate.

udif commented:
As requested, I rebased on top of develop, (fix a few bugs on the way), and added a screenshot.
As it currently stands, I can change the Spinner's state, but it is saved only if I modify some other settings causing the FeedPreferences to be saved.
Tomorrow I hope to do actual functionality checks, as I couldn't check those until now because the preferences weren't saved.

![screen shot 2015-06-28 at 1 35 35 am](https://cloud.githubusercontent.com/assets/809640/8394209/29f72eb2-1d36-11e5-95c5-5f174c127fb1.png)

udif commented:
DeleteAfterPlayback flag now seems to be saved correctly.

mfietz commented on app/src/main/java/de/danoeh/antennapod/fragment/NewEpisodesFragment.java:
```diff
@@ -97,7 +97,7 @@ public void onHide(FeedItemUndoToken token) {
                     FeedItem item = DBReader.getFeedItem(context, itemId);
                     FeedMedia media = item.getMedia();
-                    if(media != null && media.hasAlmostEnded() && UserPreferences.isAutoDelete()) {
+                    if(media != null && media.hasAlmostEnded() && item.getFeed().getPreferences().getCurrentAutoDelete(UserPreferences.isAutoDelete())) {
```
looks rather strange.
What about removing the ``isAutoDelete`` parameter and using 
```
case DEFAULT:
                return UserPreferences.isAutoDelete();
```
?

udif replied:
The reason I haven't done this in the first place is that currently FeedPreferences does not import UserPreferences, while NewEpisodeFragment and all the other files using getCurrrentAutoDelete() already does, so it looked cleaner not to add extra dependency.

The advantage of your suggestion is that it saves adding this parameter in at least 5 places where getCurrrentAutoDelete() is called.
You could say that since this function has no use other than with UserPreferences.isAutoDelete() as a parameter, then maybe it is not worth keeping the parameter.

And now a side git question:
Lets say I fix this,should I squash this commit into my pull request commit, or leave it as another commit on top of the current PR?

mfietz commented:
Looks really good, time for some nit-picking ^^
This will also override the smart delete preferences, so delete after playback might be rather confusing. 

Maybe "Auto Delete Episodes:" with "Global", "True"(?) and "False"(?)

udif commented:
OK, "Auto Delete Episode" it is :-)

udif commented:
Changed All occurrences of 'Default' into 'Global' (both strings and enums).
Changed text description from 'Delete After Playback' -> 'Auto Delete Episode'

udif commented:
@mfietz , I finished testing this and it seems to work for me, and ready to be pulled in, unless you want me to remove the parameter to isAutoDelete(), as you commented in https://github.com/AntennaPod/AntennaPod/pull/960#discussion_r33427368 , it's your call.

mfietz commented on core/src/main/java/de/danoeh/antennapod/core/storage/PodDBAdapter.java:
```diff
@@ -188,7 +189,8 @@
             + KEY_HIDE + " TEXT,"
-            + KEY_LAST_UPDATE_FAILED + " INTEGER DEFAULT 0)";
+            + KEY_LAST_UPDATE_FAILED + " INTEGER DEFAULT 0,"
+            + KEY_AUTO_DELETE_ACTION + " INTEGER DEFAULT 0)";
```
You do not upgrade the existing table, do you? Have a look at StorageCallbacksImpl.java (increment the version, ``if(oldVersion <= 16) { ...`` and add the column)
Everything looks good and ready to be pulled, but this addition is necessary

mfietz commented on core/src/main/res/values/arrays.xml:
```diff
@@ -1,6 +1,12 @@
+    <string-array name="spnAutoDeleteItems">
+        <item>Global</item>
+        <item>Always</item>
+        <item>Never</item>
```
Please use ``@string/...`` so that the values can be localized

udif replied:
Sure, I'll do that.
Out of curiosity, was it possible to simply build a tree of localized arrays.xml instead?
I guess since 90% of array.xml doesn't need to be localized, it was chosen to put all the localization in one file?
https://developer.sinnerschrader-mobile.com/moving-localization-from-arrays-xml-to-strings-xml/463/

mfietz replied:
I guess it would. The translators would see another resource and might get a bit confused why these strings are separated from the others.
I'm not even sure if Transifex understands this particular format...

mfietz commented on app/src/main/res/layout/feedinfo.xml:
```diff
@@ -159,7 +159,38 @@
+                    android:spinnerMode="dropdown"
+                    app:layout_gravity="center"
+                    android:dropDownWidth="wrap_content"
+                    android:clickable="false" />
```
Why is clickable set to false?
When testing this PR on virtual devices, the drop down menu would not open, unless I set clickable to true?

udif replied:
When I test this PR on a Virtual "Nexus 5 API 21 x86" it definitely opens the spinner.
I've checked all the Spinner combinations (Global/Yes/No) against the Auto Delete Settings both in the checked and unchecked position.

Spinner	Settings	Result
Global	Checked	Deleted
Yes		Checked	Deleted
No		Checked	Not Deleted
Global	UnChecked	Not Deleted
Yes		UnChecked	Deleted
No		UnChecked	Not Deleted

As I wrote earlier, this is the first Android project I'm involved in, and most of the coding is based on cloning existing style in AntennaPod or on googling (thanks stackexchange & github...).
I think that android:clickable doesn't matter here because I use OnItemSelectedListener().

mfietz replied:
Yeah, the listener is called once I *selected* an item from the list.
I ran two different virtual devices on Genymotion, one was Androdi 2.3, the other 4.4. In both cases, the spinner would not react to my clicks, no dropdown menu.
When I set clickable to true, at least the dropdown menu was showing. I did not do any additional testing, so this change might cause other problems.

udif replied:
What is the minimum API level /Android version I need to test against?
What VM configurations do you recommend for testing?

mfietz replied:
> What is the minimum API level /Android version I need to test against?

API Version 10 which is Android version 2.3.3 - 2.3.7

> What VM configurations do you recommend for testing?

I usually do not change the defaults.

udif commented:
I fixed the spinner strings to be taken from the resource, and also fixed the database
(Although strangely enough, when running it after the fix with my previous test database, it crashed due to duplicate column, which indicates that it already added the new column by itself??).
After I uninstalled the program completely and reran it, it was fine.

Next will be testing on a VM with 2.3.x to see how the spinner behaves there.

I also tested on a real device (Nexus 5) since the Android Emulator doesn't support Hebrew under OSX (according to a stackexchange comment, there is no such problem under Windows).

I kept today's changes on a separate branch on my home repository. Would you like me to update the PR with separate commit so that you can see more clearly just the new fixes, or would you prefer to squash all of those into a single commit (replace the existing PR with a fixed one)?

I would also like to get a final decision regarding getCurrrentAutoDelete(). I can remove the parameter, but then FeedPreferences would also need to import UserPreferences.

mfietz commented:
> Would you like me to update the PR with separate commit so that you can see more clearly just the new fixes, or would you prefer to squash all of those into a single commit (replace the existing PR with a fixed one)?

I'd prefer a separate commit, but a squashed commit is also fine.

> I would also like to get a final decision regarding getCurrrentAutoDelete()

Calling ``UserPreferences.isAutoDelete()`` from ``getCurrrentAutoDelete()`` instead of passing it as a parameter seems a lot cleaner to me. But this is only my personal preference.
But if you think it makes sense to pass is as a paremeter (perhaps getting the value from somewhere else in a future update), I won't stand in your way ;)

udif commented:
OK, I fixed all the pending issues:

Migrating database to version 17
Moved isAutoDelete() into getCurrentAutoDelete()
Android 2.3.x support (clickable:true) - I checked myself, and indeed this is needed.
Spinner strings taken from the string resource.
I also rebased the code, based on current HEAD of develop.

The pull request above includes the original pull request (rebased), plus additional commits, one for each issue.

mfietz commented:
Honestly, I don't know why but for some reason your code breaks several tests in ``PlaybackServiceMediaPlayerTest``. Trying to figure out why, but I'm pretty clueless yet

mfietz commented:
Please add 
```
FeedPreferences prefs = new FeedPreferences(f.getId(), false, FeedPreferences.AutoDeleteAction.NO, null, null);
f.setPreferences(prefs);
```
after line 120
```
Feed f = new Feed(0, new Date(), "f", "l", "d", null, null, null, null, "i", null, null, "l", false);
```
in ``PlaybackServiceMediaPlayerTest.java``

udif commented:
Are the errors below related?
I tried building and testing according to the wiki page, and I get this while running the assembleDebug task in gradle:

Note: /Volumes/Untitled/StudioProjects/AntennaPod/app/src/androidTest/java/de/test/antennapod/service/playback/PlaybackServiceMediaPlayerTest.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: /Volumes/Untitled/StudioProjects/AntennaPod/app/src/androidTest/java/de/test/antennapod/ui/MainActivityTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

udif commented:
I assume I also need to add this?

import de.danoeh.antennapod.core.feed.FeedPreferences;

mfietz commented:
Of course. My IDE pretty much manages the imports by itself, I do not really thing about them most of the time. Doesn't your IDE also do auto import?

udif commented:
Hmm... I didn't know this was even possible, or even thought about looking for such a feature!

I use AndroidStudio on OSX, and apparently it can do this...
http://stackoverflow.com/questions/16615038/what-is-the-shortcut-to-auto-import-all-in-android-studio
I learn new stuff every day!

udif commented:
OK, I ran and tested on a 2.3.3 VM (API 10) with an en_US locale, and this is what I got from ./gradlew  connectedAndroidTest

de.test.antennapod.ui.MainActivityTest > testAddFeed[Nexus_S_API_10_arm(AVD) - 2.3.3] FAILED 
	junit.framework.AssertionFailedError: Text string: 'Add Podcast' is not found!
	at com.robotium.solo.Clicker.clickOnText(Clicker.java:447)

de.test.antennapod.ui.MainActivityTest > testDrawerPreferencesHideAllElements[Nexus_S_API_10_arm(AVD) - 2.3.3] FAILED 
	junit.framework.AssertionFailedError: Text string: 'New Episodes' is not found!
	at com.robotium.solo.Clicker.clickOnText(Clicker.java:447)

de.test.antennapod.ui.PreferencesTest > testAbout[Nexus_S_API_10_arm(AVD) - 2.3.3] FAILED 
Test failed to run to completion. Reason: 'Instrumentation run failed due to 'java.lang.NoClassDefFoundError''. Check device logcat for details
Tests on Nexus_S_API_10_arm(AVD) - 2.3.3 failed: Instrumentation run failed due to 'java.lang.NoClassDefFoundError'
:app:connectedAndroidTestDebug FAILED

FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':app:connectedAndroidTestDebug'.
> There were failing tests. See the report at: file:///Volumes/Untitled/StudioProjects/AntennaPod/app/build/reports/androidTests/connected/index.html


When looking at the HTML, I see that 3 test failed, 2 of them due to the "text string not found" errors shown above, and one due to the NoClassDefFoundError.

mfietz commented:
Go to the Preferences, Editor, General, Auto Import, check *Optimize imports on the fly* and *Add unambiguous imports on the fly*
That should cover most cases

udif commented:
Yes, I already enabled the auto import stuff after searching for this and finding the stackoverflow link. Neat!

mfietz commented:
*testAbout* failing is "normal" - try running testDrawerPreferencesHideAllElements again

udif commented:
How do I run an individual test?
so far I simply ran:
./gradlew connectedAndroidTest

mfietz commented:
I run the tests from Android Studio exclusively. Why would you even use the console? ;)

udif commented:
I ran them from the command line because I didn't know how to do this from AndroidStudio :-)

Now I know it's possible, is it done by right mouse click on MainActivityTest, then selecting Run -> MainActivityTest ?

mfietz commented:
You probably tried it out by the time I answer this... Yes, it is...
Also, there are a lot of blogs and alike that show how to do particular stuff in AS. Googling it should not be too hard

TomHennen commented:
I run them from the command line. Not sure I trust AndroidStudio all the
time.

On Tue, Jul 7, 2015 at 6:30 PM Martin Fietz <notifications@github.com>
wrote:

> You probably tried it out by the time I answer this... Yes, it is...
> Also, there are a lot of blogs and alike that show how to do particular
> stuff in AS. Googling it should not be too hard
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AntennaPod/AntennaPod/pull/960#issuecomment-119363384>
> .
>

mfietz commented:
Aaaaand this needs a rebase

udif commented:
@mfietz Please check the last commit (Fix DB to version 18). The rest are simple rebase with no collisions.

@TomHennen I'm having a hard time getting all tests to pass even without my PR. Can you please provide me with a known point where I can see all 117 tests pass? Bt that I mean the exact commit (SHA1) from which you tested, and the platform on which you ran it (which Android version , which emulator, etc.). I tried checking out from the top of upstream/develop and run all tests, but now I get 14 failed tests. I use the latest AndroidStudio on OSX.

TomHennen commented:
I have good results with the 'Google Nexus S - 4.1.1 API 16" emulator in
Genymotion.  The current develop should work just fine.

On Wed, Jul 8, 2015 at 7:12 PM udif <notifications@github.com> wrote:

> @mfietz <https://github.com/mfietz> Please check the last commit (Fix DB
> to version 18). The rest are simple rebase with no collisions.
>
> @TomHennen <https://github.com/TomHennen> I'm having a hard time getting
> all tests to pass even without my PR. Can you please provide me with a
> known point where I can see all 117 tests pass? Bt that I mean the exact
> commit (SHA1) from which you tested, and the platform on which you ran it
> (which Android version , which emulator, etc.). I tried checking out from
> the top of upstream/develop and run all tests, but now I get 14 failed
> tests. I use the latest AndroidStudio on OSX.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AntennaPod/AntennaPod/pull/960#issuecomment-119755959>
> .
>

udif commented:
OK, all tests now pass using Genimotion and Android 5.1.0 image.
Earlier I stashed @mfietz code snippet that fixed one of the test files, and then forgot to pop it back.

udif commented:
As much as I should be happy about the merge, I reran the tests against a Genimotion Nexus 4 v4.2.2 , and I see some failures (16).
All errors are from PlaybackServiceMediaPlayerTest.
Most of the tests fail due to "Unexpected state change: ERROR", but there are also a few 

I've also tested on a Genimotions Nexus S 2.3.7. I got one error on testAbout():
Test failed to run to completion. Reason: 'Instrumentation run failed due to 'java.lang.NoClassDefFoundError''. Check device logcat for details

I've looked at logcat and couldn't find anything unusual.

mfietz commented:
> I've also tested on a Genimotions Nexus S 2.3.7. I got one error on testAbout():

As I already explained, this is normal. On 2.3.7, the app uses ``PreferenceActivityGingerbread`` instead of ``PreferenceActivity``. Basically, that test should not be run on 2.3.x, but as so often: not enough time to work around this, not really believing this is worth the effort...

mfietz commented:
Just re-run all of ``PlaybackServiceMediaPlayerTest``'s tests on a 4.4.4 virtual device, all of them passed without any problem.

TomHennen commented:
I ran all the tests on a 4.1.1 system. Everything passed without a problem.  Our tests do sometimes have hiccups on different emulators. I haven't had time to look in to this fully, but I suspect it's mostly timing problems.
